### DISCUSSION TITLE
How to approach modularity?

### DISCUSSION THREAD

**Discussion by ghost** [littleK0i/SnowDDL#70] [Q&A] [answered]
Suppose that I want 300 instances of a custom abstraction called "source." 

Let's say that each source contains:
* qty 1 - FILE FORMAT object
* qty 1 - STAGE object
* qty 1 - TABLE object
* qty 1 - PIPE object

How can I make a "template" of this group of resources and make parameterized instances of the group without doing it as raw resource instances?

Thanks!

**Comment by littleK0i** [comment, marked as answer]
Python itself is the best template engine with 100% flexibility.

You may use "programmatic config" feature to generate objects dynamically based on your custom logic.
https://docs.snowddl.com/advanced/programmatic-config

1. Create directory `__custom` (with two leading underscores).
2. Put a python module with method "handler" into `__custom` directory.
3. Write code to build or modify whatever you want.

```
from snowddl import *

def handler(config: SnowDDLConfig):
    file_format_bp = FileFormatBlueprint(...)
    stage_bp = StageBlueprint(...)
    table_bp = TableBlueprint(...)
    pipe_bp = PipeBlueprint(...)

    ...

    config.add_blueprint(file_format_bp)
    config.add_blueprint(stage_bp)
    config.add_blueprint(table_bp)
    config.add_blueprint(pipe_bp)
```

You may read "sources" from some other config or external endpoint. You may run loops and do anything you would normally do in Python. You may replace existing blueprints with updated versions. Etc. etc.

**Reply by ghost** [reply]
Thank you! This helped greatly.

I have been playing with some code, and realized that I would like a YAML config file for each "source" as if it were a real SnowDDL/snowflake resource. Looking at the code for those, I see they use a Parser to achieve behavior similar to what you have above, but for single resources.

ref: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/parser/file_format.py

The `__custom` examples out there so far do not read from such YAML files out of the box. I can certainly write my own code to fetch the yaml from files like `__custom/source/foo.yaml` where `__custom/01_source.py` is the handler. I decided to post though to ask if you see an opportunity for reusing the AbstractParser for my use case. I have defined a JSON schema object for my yaml files and I added static blueprints for the four resources (file format, stage, table, pipe). Do you see a slick way to pull in the YAML for me? Thank you!

P.S. If we achieved this, we'd effectively have the ability to add our own custom SnowDDL object support w/SnowDDL-style YAML at user-side.  I suppose we already had that, but it's a bit DIY atm.

**Reply by ghost** [reply]
I think this would be key for me to use: https://github.com/littleK0i/SnowDDL/blob/master/snowddl/parser/_parsed_file.py

**Reply by littleK0i** [reply]
It is possible.

1. Create your own "Parser" object and derive it from `AbstractParser` class.
2. Create your own "App" object by overloading "BaseApp". There is a property called `parser_sequence`, which you can extend and add your own parser. https://github.com/littleK0i/SnowDDL/blob/master/snowddl/app/base.py#L24
3. Run overloaded app by calling `.execute()` method. https://github.com/littleK0i/SnowDDL/blob/master/snowddl/app/base.py#L528-L530

Entry-points are currently configured in `setup.cfg` and created during module installation. But you may write your own Python script and call it instead of `snowddl` CLI command.

One parser can add any number of blueprints for any object types.
